compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlCodeLowering.cpp

```diff
@@ -23,34 +23,61 @@ struct HalfDmaCpyNdToNpuConverter final
   using OpConversionPattern::OpConversionPattern;
 
   HalfDmaCpyNdToNpuConverter(MLIRContext *context,
-                             const AMDAIE::AMDAIEDeviceModel &deviceModel)
-      : OpConversionPattern(context), deviceModel(std::move(deviceModel)) {
+                             const AMDAIE::AMDAIEDeviceModel &deviceModel,
+                             bool lowerCtrlpktDma)
+      : OpConversionPattern(context),
+        deviceModel(std::move(deviceModel)),
+        lowerCtrlpktDma(lowerCtrlpktDma) {
     minStrideBitWidth = deviceModel.getMinStrideBitWidth();
   }
 
   /// Insert ops to write a BD, patch the address and push it to the queue. This
   /// is specific to Shim BDs for now.
   FailureOr<AMDAIE::NpuPushToQueueOp> insertWriteBdOps(
       AMDAIE::NpuHalfDmaCpyNdOp op, ConversionPatternRewriter &rewriter,
-      AMDAIE::AMDAIETileType tileType,
-      AMDAIE::LogicalObjectFifoFromMemrefOp logicalObjFifo,
-      AMDAIE::BdIdOp bdIdOp, AMDAIE::ChannelOp channelOp, int64_t bufferLength,
-      int64_t bufferOffset, int32_t enablePacket, int32_t packetId,
-      int32_t packetType, SmallVector<OpFoldResult> sizes,
+      AMDAIE::AMDAIETileType tileType, AMDAIE::BdIdOp bdIdOp,
+      AMDAIE::ChannelOp channelOp, int64_t bufferLength, int64_t bufferOffset,
+      int32_t enablePacket, int32_t packetId, int32_t packetType,
+      SmallVector<OpFoldResult> sizes,
       SmallVector<OpFoldResult> strides) const {
     uint8_t numIntraAddrDim = deviceModel.getDmaProp<uint8_t>(
         tileType, AMDAIE::AMDAIEDmaProp::NumAddrDim);
     uint8_t numAddrDim =
         numIntraAddrDim + deviceModel.deviceConfig.dmaNbInterDims;
-    auto subspanOp = dyn_cast_if_present<IREE::HAL::InterfaceBindingSubspanOp>(
-        logicalObjFifo.getMemref().getDefiningOp());
-    if (!subspanOp) {
-      return logicalObjFifo.emitOpError()
-             << "must operate on an `hal.interface.binding.subspan`";
+
+    int64_t argIdx;
+    int64_t elemWidthInBits;
+    uint8_t memSpace;
+    if (lowerCtrlpktDma) {
+      // Control packet DMAs require special handling.
+      // The `argIdx` must match the driver and is assumed to be 0.
+      // The `elemWidthInBits` is fixed at 32 for control packet data.
+      // The `memSpace` is set to 0, indicating storage in global memory.
+      argIdx = 0;
```

Comment:
Is this reserving kernel argument 0 for control packets? But the normal data buffers are also offset from zero, so I think you're assuming here that there will be 2 runs? One to reconfigure and one to perform the actual operation?

Reply:
Thanks, makes sense